feat(ci-labware): Labware deploy workflow

neo-jesse opened the pull request:
# Overview

This PR creates a deploy workflow for the Labware Library. It is very similar to the now-existing MKDOCS workflow.

## Test Plan and Hands on Testing

Testing this PR includes using the temp tags "tmp-labware-library-*" and "tmp-staging-labware-library-*" to deploy to prod and stag, respectively. 

When we are ready, the PR needs to be updated with the 'edge' branch and the 'tmp-' part of the tag needs to be removed.

smb2268 committed:
fix(components): Render waste chute fixture in stacker D3 location to preserve ordering (#19443)

Fix [RQA-4634](https://opentrons.atlassian.net/browse/RQA-4634)

jerader committed:
fix(protocol-designer): adding a staging area slot retains previous s… (#19447)

…taging areas

closes RQA-4636

emilyburghardt committed:
docs(api): order of operations in a liquid class transfer (#19402)

<!--
Thanks for taking the time to open a Pull Request (PR)! Please make sure
you've read the "Opening Pull Requests" section of our Contributing
Guide:


https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

GitHub provides robust markdown to format your PR. Links, diagrams,
pictures, and videos along with text formatting make it possible to
create a rich and informative PR. For more information on GitHub
markdown, see:


https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

To ensure your code is reviewed quickly and thoroughly, please fill out
the sections below to the best of your ability!
-->

# Overview

Updating the order of operations section to be more accurate for _every
action enabled_. I used the info from David and Jeremy and the liquid
classes flow chart Jeremy/Sanniti and I developed before liquid classes
release.
<!--
Describe your PR at a high level. State acceptance criteria and how this
PR fits into other work. Link issues, PRs, and other relevant resources.
-->

## Test Plan and Hands on Testing

<!--
Describe your testing of the PR. Emphasize testing not reflected in the
code. Attach protocols, logs, screenshots and any other assets that
support your testing.
-->

## Changelog

- added a note underneath the order of operations list to capture the
differences between 1) blowout at destination well and 2) blowout at
other location. I wanted to avoid a branching list and keep it clean.
<!--
List changes introduced by this PR considering future developers and the
end user. Give careful thought and clear documentation to breaking
changes.
-->

## Review requests

<!--
- What do you need from reviewers to feel confident this PR is ready to
merge?
- Ask questions.
-->
Check for accuracy.

## Risk assessment

<!--
- Indicate the level of attention this PR needs.
- Provide context to guide reviewers.
- Discuss trade-offs, coupling, and side effects.
- Look for the possibility, even if you think it's small, that your
change may affect some other part of the system.
- For instance, changing return tip behavior may also change the
behavior of labware calibration.
- How do your unit tests and on hands on testing mitigate this PR's
risks and the risk of future regressions?
- Especially in high risk PRs, explain how you know your testing is
enough.
-->
Low.

ddcc4 committed:
test(api): silence the `func-returns-value not covered` lint warning (#19449)

# Overview

lint-py is complaining that:
```
tests/opentrons/protocol_api/core/engine/test_transfer_components_executor.py:1924: note: Error code "func-returns-value" not covered by "type: ignore" comment
```

I had an `ignore[call-arg]` there because I just wanted to spy on how
many times `mock_instrument_core.touch_tip()` was called without
specifying the arguments (PR #19425). But the linter also wants me to
add an `ignore[func-returns-value]` there.

## Test Plan and Hands on Testing

Ran `make lint-py` again.

## Risk assessment

Low. No change to code or functionality at all. Just tweaking a lint
warning.

smb2268 committed:
fix(app): hide module banners if subsystem update is ongoing (#19450)

Fix RQA-4635

sfoster1 committed:
feat(api): add a leak check analysis cli (#19452)

We've realized that it's actually pretty easy for protocols with complex
code and functionality to accidentally leak a protocol engine. The long
term fix for this is to get protocols running in their own processes, or
at the very least their own interpreters; for now though, what we can do
is add a flag to the analysis cli that checks for the presence of such a
leak.

## Testing
- [x] leak an engine in a protocol and check that this prints it (I can
provide one that does this)
- [x] have a protocol that doesn't and see that it doens't print that

We should probably get this running automatically on our test harness.

---------

Co-authored-by: Josh McVey <josh.mcvey@opentrons.com>

jbleon95 committed:
fix(api): replace labware ID in error messages with user readable names (#19455)

Removes printing the raw labware ID in several error messages, replacing them with the display name/label or load name.

jerader committed:
refactor(protocol-designer): fix styling and copy of release notes (#19457)

modify release notes copy and formatting

koji committed:
fix(app-shell): change ubuntu version in Makefile from 22 to 24 (#19445)

* fix(app-shell): change ubuntu version in Makefile from 22 to 24

ddcc4 committed:
test(api): I give up, spell out mock touch_tip() arguments (#19453)

# Overview

I thought I could make the `test_transfer_components_executor.py` more
succinct by leaving out the arguments to the mock `touch_tip()` and
doing:
```
decoy.verify(
    mock_instrument_core.touch_tip(),  # type: ignore[call-arg]
    ignore_extra_args=True,
)
```
but lint didn't like that.

I fixed the lint warning, but then @jbleon95 discovered that pytest
itself complains with the warning:
```
IncorrectCallWarning: missing a required argument: 'location'
```
The warning doesn't fail the test, but it does add a bit of noise to the
logs.

To keep the test logs clean, I gave in and spelled out all the arguments
to `touch_tip()`.

## Test Plan and Hands on Testing

Ran `test_transfer_components_executor.py` with warnings turned on.

## Risk assessment

Low, cosmetic change to test.

neo-jesse committed:
first run at deploy

koji committed:
chore(protocol-designer): modify init config (#19444)

*chore(protocol-designer): modify init config

neo-jesse committed:
more pipeline updates

ryanthecoder committed:
feat(api): connect TC ramp rate from protocol api through protocol engine (#19451)

# Overview
First of several small PR's aimed at connected the thermocyclers ramp
rate control from protocol api through firmware.

This PR connects the already existing protocol api argument to the
protocol engine command.
This doesn't add any new features yet but just mends the first break in
the chain so there's nothing to test yet outside of the unit tests

---------

Co-authored-by: Max Marrone <max@opentrons.com>

neo-jesse committed:
tmp tags

Commit:
fix(opentrons-ai-client): fixing image path and logo issue  (#19459)

# Overview
Closes AUTH-2258

This PR fixes missing labware images issue and logo issue.

**Logo issue**
- `rem` was used instead of `px` -- now changed to `px`

<img width="713" height="117" alt="image"
src="https://github.com/user-attachments/assets/c856689c-6d73-4576-9611-74a39303ff01"
/>



 **Image Issue**:

7 labware images failed to load on staging/prod (worked locally). The
images used new URL() with import.meta.url but had incorrect paths.

  **Root Cause:**

new URL('/ai-client/assets/...', import.meta.url) doesn't work correctly
with Vite's production builds - the /ai-client/ alias isn't resolved
properly, causing 404s.

  **Solution:**
  Converted from new URL() dynamic imports to direct ES6 imports:

  // ❌ Before (broken in production)
  new URL('/ai-client/assets/images/labware/image.jpg',
  import.meta.url).href

  // ✅ After (works in production)
  import image from '/ai-client/assets/images/labware/image.jpg'

Files changed: Renamed /labwares/ → /labware/ (consistency) + converted
  to direct imports in labware-images.ts

<img width="823" height="121" alt="image"
src="https://github.com/user-attachments/assets/c6b9a0c3-2fcd-4296-8737-3df08b3c6f71"
/>




## Test Plan and Hands on Testing

Please run and see if labware images are shown in Labware & Liquids
section.



## Risk assessment

Mid

y3rsh commented on .github/workflows/labware-build-deploy.yaml:
We won't need this new workflow.
The existing workflow, the one in #19469, is the one we want to alter.
 
We want to add to that workflow

- [ ] skipping tests if trigger is a tag
- [ ] also deploying to the new sandbox (deploy to the tag instead of branch if tag trigger)
- [ ] if the right kind of tag deploying to the new staging and production

y3rsh commented on labware-library/Makefile:
```diff
@@ -68,3 +72,20 @@ test:
+
+.PHONY: deploy-production
+deploy-production:
+	$(MAKE) deploy ENVIRONMENT=production
```
I think we don't want to alter this Makefile
Let us instead add `deploy:ci` and `deploy:local` to `scripts/static-deploy/Makefile`

y3rsh commented on scripts/static-deploy/labware_deploy_config.py:
This logic all belongs in `scripts/static-deploy/deploy_config.py`

That script will parse the tag and the other parameters to determine the application and environment.  
Then call `get_config(environment, application)` to get the DeploymentConfig
This centralizes all logic for config mapping.  I can then look in that file and see 

- exactly the tags that matter for each application
- exactly the AWS resources that map to each environment

y3rsh commented on scripts/static-deploy/test_deploy_labware.py:
this should go in the `tests` directory

y3rsh commented on scripts/static-deploy/deploy_labware.py:
This script will be called deploy.

Its job is to consume a configuration object and artifact_root that specifies the application and all the details of the AWS resources for that application.

I know we originally said the AWS CLI would be fine since it’s always available on GitHub runners, but after thinking it through I’d actually prefer we use **Boto3** here.

**Reasons:**
- **Pinned dependency**: We will lock Boto3 in `uv`, ensuring consistent versions across environments. With the CLI, we’re tied to whatever the GitHub runner ships.
- **Cleaner error handling**: Using Boto3 lets us catch and handle exceptions in Python directly, rather than parsing stdout/stderr from a subprocess call.
- **Consistency**: Keeps all logic in Python instead of mixing shell commands with Python. Easier to test, log, and maintain.

Overall, Boto3 gives us better control and consistency at basically no extra cost. I think we should add it to our `uv` setup and make it the default here.

in the `scripts/static-deploy` directory 
run
`uv add boto3`

y3rsh reviewed (commented):
Going to try and be clearer about the vision and requirements of the flows

# Deploy Requirements (Local vs CI)

## Summary
We support **two explicit entry points**:
- `deploy:ci` — used in GitHub Actions; determines env/app from the triggering event/tag.
- `deploy:local` — used by humans; requires an explicit override of the application and environment.  In the case of the sandbox environment we also want an override of `folder` or some other better name that indicates the prefix of this deployment.

Both entry points produce the **same configuration object** and call the same deploy function. The only difference is **how inputs are resolved**.

---

## Invariants (apply to both flows)
- A single **configuration resolution layer** produces a normalized `DeployConfig`:
  - `application` (e.g., `labware-library`, `protocol-designer`, `makedocs`, `ai-client`) - for this PR on need `labware-library` mapped and working 
  - `environment` (e.g., `sandbox`,`staging`, `prod`, )
    - get_config(environment, application) maps to the data we need for deployment, like;
      - `distribution_id` (CloudFront)
      - `bucket` (target S3 bucket)
      - and so on
- Strong validation: fail early if any required field is missing/contradictory.

---

## `deploy:ci` (CI Flow)
**Purpose:** Fully automated deploy from a tag/PR merge.

**Inputs (implicit/derived):**
- Git metadata (e.g., `GITHUB_REF`, `GITHUB_EVENT_NAME`, and such) Tag/branch naming convention → maps to `{application, environment}`
- CI-provided artifact location (downloaded or produced earlier in the workflow)
- no profile is specified because `aws-actions/configure-aws-credentials@v4` sets it in the environment 

```python
session = boto3.Session(profile_name=profile) if profile else boto3.Session()
s3_client = session.client("s3", region_name=config.region)
```

**Resolution rules:**
1. Parse ref and workflow to infer `environment` + `application` .

**Validation:**
- Ensure derived `artifact_root` exists.
- Ensure app/env is resolved correctly
- Print a redacted config summary before execution.

**Flow**
1. a function that intakes all the info about the github runner environment and trigger and determines environment and application.
2. Resolve the config - this function already exists `get_config(environment, application,)`.
3. call `deploy(config, artifact_root)`.

---

## `deploy:local` (Manual Flow)
**Purpose:** Human-operated deploy when GitHub Actions is down or for controlled manual releases.

**Preconditions:**
- The local user has **already built** the app
- The user will be passing a profile

**Inputs (explicit):**
- `--application <name>`
- `--environment <name>`
- `--artifact-root <path>`
- `--profile <string>`

**Resolution rules:**
1. Validate `artifact_root` exists and is readable.

**Flow**
1. Resolve the config - this function already exists `get_config(environment, application,)`.
2. call `deploy(config, artifact_root, profile)`.

mjhuff committed:
feat(app, app-shell): support secondary windows in the desktop app (#19461)

Works toward EXEC-1760

With the camera enablement initiative, Design is implementing a new pattern utilizing "secondary windows." These are browser windows independent of the main browser window. We'll be using these for several items during camera enablement, such as the live stream and camera images, and because we will use these windows in the future, it makes sense to build a scalable architecture for adding secondary windows.

This commit adds the app-shell architecture for secondary windows. Secondary windows effectively have a "type", which is loosely associated to the kind of content the window renders. Each type of secondary window emits an action, and from that payload we build an id to determine the uniqueness of the content. If the content is unique, we open a new window, otherwise we refocus the window already displaying that content.

Because secondary windows do not require access to all action handlers, we only pass in a subset of them. Ex, we don't want update prompting to occur in a secondary window, so we don't need to supply the update handler to secondary windows.

The camera stream wire-up is added here for completeness and to provide a working example of the general interface.

neo-jesse committed:
making suggested changes

jerader committed:
feat(step-generation): add lid_namespace and lid_version to loadLabware (#19470)

closes AUTH-2224

neo-jesse committed:
DOCS_SANDBOX_ROLE

neo-jesse committed:
testing speed improvement

neo-jesse committed:
OIDC write issue

jerader committed:
feat(step-generation): bump to 2.26 api version for emitted protocols (#19471)

neo-jesse committed:
boto3

neo-jesse committed:
Fix labware deploy workflow: add UV setup and remove non-tmp tag patterns

neo-jesse committed:
Resolve merge conflicts: use shared JS setup action, keep new deploy configuration

neo-jesse committed:
testing deploy to sandbox

neo-jesse committed:
Add debug step to determine-deploy-config job to troubleshoot staging tag deployment

neo-jesse committed:
Add debug steps to build-ll and deploy-ll jobs to troubleshoot job execution

neo-jesse committed:
Add explicit job dependency checks and debug step to identify why deploy-ll job is skipped

neo-jesse committed:
Fix job dependencies: make build-ll wait for both js-unit-test and e2e-test to complete
